adds blog post _The secret life of data in Valkey_ (#575)
valkey-io/valkey-io.github.io: content/blog/2026-06-24-secret-life-of-data/index.md

+++
title = "The secret life of data in Valkey"
date = 2026-06-24

> description = "If you’ve used Valkey for very long you’re probably well aware of the different data types, however what lies under the surface of these types could be the missing piece of your cost optimization strategy."

Reviewer: nit: run on, punctuation

```suggestion
description = "If you’ve used Valkey for a long time, you’re probably familiar with its different data types. However, what lies under the surface of these types could be the missing piece of your cost optimization strategy."
```

authors = ["kyledvs"]
[taxonomies]
blog_type = ["Technical Deep Dive"]
[extra]
featured = true
featured_image = "/assets/media/featured/random-05.webp"
+++

> If you’ve used Valkey for very long you’re probably aware of the different primary data types in the core: strings, hashes, lists, sets, sorted sets, and streams. Each data type has an accompanying set of commands that manipulate that data type. For example, you use [`HSET`](/commands/hset/) to set a field and value pair to a hash, but you don’t use, for example, HSET to add an element to a list. What if I told you that there is an entire hidden layer that you never see unless you use a specific `OBJECT` command and can’t control unless you start tinkering with the configuration? And that this could have a profound effect on your total usable storage: enough that careful configuring could mean significant savings in your infrastructure. Read on to find out how you can understand and tweak the _real_ underlying structures.  

Reviewer: some optional revisions

```suggestion
If you’ve used Valkey for a long time, you’re probably aware of the different primary data types in the core: strings, hashes, lists, sets, sorted sets, and streams. Each data type has an accompanying set of commands that manipulate that data type. For example, you use [`HSET`](/commands/hset/) to set a field-value pair in a hash, but you wouldn't use it to add an element to a list. What if I told you that there is an entire hidden layer that you never see unless you use a specific `OBJECT` command, and that you can't control unless you start tinkering with the server configuration? This hidden layer can profoundly affect your total usable storage: enough that careful configuration could lead to significant infrastructure cost savings. Read on to find out how you can understand and tweak the _real_ underlying structures.
```

## Engineering tradeoffs

> Compact memory representation is highly important in Valkey: most users aren’t limited by performance but they are limited by the amount of memory available. When you’re representing data in RAM there often isn’t a clear winner for a technique that maximizes both performance and compactness at all data sizes. Like every engineering project, Valkey attempts to balance tradeoffs of performance and compactness.

Reviewer: optional revisions

```suggestion
Compact memory representation is crucial in Valkey; most users are limited by available memory rather than performance. When you’re representing data in RAM, there often isn’t a clear winner for a single technique that maximizes both performance and compactness at all data sizes. Like every engineering project, Valkey attempts to balance the tradeoffs between performance and compactness.
```

> In this case, the balance is achieved by an _encoding_. All data in Valkey has an encoding, and this abstracts the actual in-memory representation from how it’s accessed (i.e. the commands used). Take the example of these three HSETs (assume an empty database):

Reviewer: optional revisions

```suggestion
Valkey achieves this balance using _encodings_. Every object in Valkey has an encoding that abstracts its actual in-memory representation from how it is accessed (i.e., the commands used). Consider these three `HSET` commands in an empty database:
```

```bash
> HSET hash0 field "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
> HSET hash1 field "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
> HSET hash2 field "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
```

> At first glance you might observe that each value is one character longer than the previous. However, what you probably don’t see is that despite the single character difference, the memory used changes dramatically. You can see it by running [`MEMORY USAGE`](/commands/memory-usage/) over each key:

Reviewer: optional revision

```suggestion
At first glance, you might notice that each value is one character longer than the previous. However, despite only a single-character difference, the memory usage changes dramatically. You can observe this by running the [`MEMORY USAGE`](/commands/memory-usage/) command on each key:
```

> | Key   | Length of value | `MEMORY USAGE` | Size difference than previous |

Reviewer: nit: grammar, should be "difference from" or "change from". I also think this is sufficient:

```suggestion
| Key   | Length of value | `MEMORY USAGE` | Change from previous |
```

| ----- | --------------- | -------------- | ----------------------------- |
| `hash0` | 63 | 104 | n/a |
| `hash1` | 64 | 120 | +15.38% |
| `hash2` | 65 | 212 | +76.67% |

_Note:  Your numbers may vary slightly depending on platform and architecture_

> So, between a value length of 64 and 65 bytes what changed to account for over a 75% increase in memory usage? 

Reviewer: nit: comma

```suggestion
So, between a value length of 64 and 65 bytes, what changed to account for over a 75% increase in memory usage?
```

> Take a look at the results of the [`OBJECT ENCODING`](/commands/object-encoding/) command for the keys in question:

Reviewer: I think it flows better like this, since we have introduced encodings, but not the command yet:

```suggestion
We can check the encoding of these keys using the [`OBJECT ENCODING`](/commands/object-encoding/) command:
```

```bash
> OBJECT ENCODING hash0
"listpack"
> OBJECT ENCODING hash1
"listpack"
> OBJECT ENCODING hash2
"hashtable"
```

> The 'hashtable' result might be expected, after all you’re using a hash command, but what’s a 'listpack'? Describing the intricacies of the 'listpack' encoding isn’t the point of this post (see [the code](https://github.com/valkey-io/valkey/blob/unstable/src/listpack.c)). The point is understanding how you interact with data is abstracted from the actual way it’s stored (an encoding). Let’s explore something else:

Reviewer: some grammatical revisions

```suggestion
The `hashtable` encoding is expected; after all, you’re using a hash command. But what’s a `listpack`? Describing the intricacies of the `listpack` encoding is beyond the scope of this post (see [the code](https://github.com/valkey-io/valkey/blob/unstable/src/listpack.c)). The key takeaway is that how you interact with data is abstracted from the actual way it’s stored (an encoding). Let’s look at another example:
```

> The 'hashtable' result might be expected, after all you’re using a hash command, but what’s a 'listpack'? Describing the intricacies of the 'listpack' encoding isn’t the point of this post (see [the code](https://github.com/valkey-io/valkey/blob/unstable/src/listpack.c)). The point is understanding how you interact with data is abstracted from the actual way it’s stored (an encoding). Let’s explore something else:

Reviewer: Do you think a brief description of listpack would be helpful? 99.9% of readers are not going to the code. Also, if we are going to later tell the user to edit the configurations, setting up a mental model of what this even means is important.

e.g.

A listpack is a memory-efficient data structure that stores elements sequentially in a single contiguous chunk of memory. For small data structures, it saves memory by minimizing the overhead of pointers. See [the code](https://github.com/valkey-io/valkey/blob/unstable/src/listpack.c) to dive deeper.

Author: Added a version of that to my revision. I don't want to get too bogged down into it since I don't think how list packs works is very important for the point of this piece.

That said, I think we _should_ spend more time on this in a different piece. It's exposed enough to matter to users but we don't really explain it well anywhere.

```bash
> LPUSH list1 foo
(integer) 1
> OBJECT ENCODING list1
"listpack"
> SADD set1 "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
(integer) 1
> OBJECT ENCODING set1
"hashtable"
```

> So, we have a list key using the same encoding as a hash and a set encoded as a hashtable. [Dogs and cats living together](https://www.youtube.com/watch?v=-9NMt42il4Q): what is going on? Conventional wisdom is that Valkey stores data using the classical data structures like hash tables, linked lists, and sets based on the commands you use. 

Reviewer: ```suggestion
Here, a list key uses the same `listpack` encoding as our hash, while a set key is encoded as a `hashtable`. [Dogs and cats living together](https://www.youtube.com/watch?v=-9NMt42il4Q): what is going on? Conventional wisdom suggests that Valkey stores data using classical data structures like hash tables, linked lists, and sets based on the commands you use.
```

Turns out: no.

> In Valkey, the _data model_ is that of classical data structures but under-the-hood, Valkey is making choices for you so you only need to worry about using the right commands for the data type and not caring about the underlying encoding. Instead, Valkey sorts out the difference between the data model and the encoding. 

Reviewer: ```suggestion
While Valkey presents a _data model_ based on classical data structures, it makes optimization decisions under the hood. This allows you to use the appropriate commands for your data type without needing to worry about how that data is stored in memory. Valkey automatically manages the mapping between the data model and the encoding.
```

## Taking back control

> With the exception of strings, which use hard-coded logic, encodings for each type are dictated by one or more configuration thresholds: anything at or below a threshold uses one encoding and anything above uses a different one. Take a look at the default configuration for the hash data type in [Valkey 9.1](https://github.com/valkey-io/valkey/releases/tag/9.1.0):

Reviewer: ```suggestion
Except for strings, which use hard-coded logic, encodings for other data types are dictated by one or more configuration thresholds: values at or below a threshold use one encoding, while values exceeding it use another. For example, consider the default configuration for the hash data type in [Valkey 9.1](https://github.com/valkey-io/valkey/releases/tag/9.1.0):
```

```bash
hash-max-listpack-entries 512
hash-max-listpack-value 64
```

> This means that any hash with all values less than or equal to 64 in length (`hash-max-listpack-value`) and have less than 512 field-value pairs (`hash-max-listpack-entries`) will be encoded as a 'listpack', everything else as a 'hashtable'. In the original example, key `hash2` had a length of 65, triggering an encoding of 'hashtable', which is much less compact accounting for the 76.67% increase in size. 

Reviewer: ```suggestion
This means that a hash is encoded as a `listpack` if all its values are at most 64 bytes long (`hash-max-listpack-value`) and it contains fewer than 512 field-value pairs (`hash-max-listpack-entries`). If either threshold is exceeded, Valkey uses a `hashtable` encoding. In the original example, the value in `hash2` was 65 bytes long, triggering a transition to the `hashtable` encoding, which is significantly less compact and accounts for the 76.67% increase in memory usage.
```

There are encoding configurations for:

- Hashes (`hash-max-listpack-entries`, `hash-max-listpack-value`)
- Lists (`list-max-listpack-size`)
- Sets (`set-max-intset-entries`, `set-max-listpack-entries`, `set-max-listpack-value`)
- Sorted Sets (`zset-max-listpack-entries`, `zset-max-listpack-value`)

> Streams still use an encoding but there isn’t a threshold because there is only one encoding option. Module types also use encodings but it would vary depending on the logic and/or configuration of the module in question.

Reviewer: ```suggestion
Streams also use encodings, but they do not have configuration thresholds because they use a single encoding. Module types also use encodings, but their behavior depends on the specific module's logic and configuration.
```

### Should you change these settings?

> Maybe. It depends on what you’re optimizing for and what your data looks like. Valkey's default encoding configuration values are fine for many use cases, but if your strategy is to cost-optimize or maximize your infrastructure, it's worth at least evaluating the configuration compared to your specific data/usage.

Reviewer: ```suggestion
Maybe. It depends on what you’re optimizing for and what your data looks like. Valkey's default encoding configurations are suitable for many use cases, but if your strategy is to cost-optimize or maximize infrastructure efficiency, it's worth evaluating these settings against your specific data patterns and usage.
```

> For the most basic evaluation, you can sample keys using `OBJECT ENCODING` to identify spots where the data is using the more voluminous encoding, then understand what is triggering it and make a call on the configuration. Of course, there are no-brainer situations: let’s say you store data in hashes and they tend to have values in length of 65 bytes, right above the threshold, and you have lower throughput needs. Upping the configuration value for `hash-max-listpack-value` to match a realistic length for your data could deliver huge savings either by lowering the infrastructure requirements or [letting you cache more with the same cluster](https://www.youtube.com/watch?v=cd4-UnU8lWY). 

Reviewer: ```suggestion
For a basic evaluation, you can sample keys using `OBJECT ENCODING` to identify where data uses a less compact encoding, determine what triggered the transition, and adjust the configuration accordingly. There are clear-cut cases: for example, if your hashes consistently contain values that are 65 bytes long (just above the default threshold) and your application has moderate throughput requirements, increasing the configuration value for `hash-max-listpack-value` to accommodate your data can yield significant savings by either reducing infrastructure requirements or allowing you to [store more data on the same cluster](https://www.youtube.com/watch?v=cd4-UnU8lWY).
```

Author: Original: 'you have lower throughput needs' 
Revision 'your application has moderate throughput requirements' 

:| 

The meaning has changed.

> For the most basic evaluation, you can sample keys using `OBJECT ENCODING` to identify spots where the data is using the more voluminous encoding, then understand what is triggering it and make a call on the configuration. Of course, there are no-brainer situations: let’s say you store data in hashes and they tend to have values in length of 65 bytes, right above the threshold, and you have lower throughput needs. Upping the configuration value for `hash-max-listpack-value` to match a realistic length for your data could deliver huge savings either by lowering the infrastructure requirements or [letting you cache more with the same cluster](https://www.youtube.com/watch?v=cd4-UnU8lWY). 

Reviewer: Probably a good callout to include

> Contextualizing an extreme example: having a 100gb cluster (5 nodes, 20gb each) packed with 95% of the keys reaching 1 over the `hash-max-listpack-value`. 

Reviewer: ```suggestion
To illustrate, let's look at an extreme example: a 100 GB cluster (5 nodes, 20 GB each) where 95% of the keys exceed the `hash-max-listpack-value` threshold by just 1 byte (65 bytes total).
```

|          | Cluster size | gb wrong-sized  | Wrong-sized key size            | Primary size | # of Primaries |
| -------- | ------------ | --------------- | ------------------------------- | ------------ | -------------- |
| Original | 100gb        | 95gb            |  212                            | 20gb        | 5              |

> | New      | 58.58gb      | 0               |  120 <br /> (56.6% of original) | 20gb        | 3              |

Reviewer: nit: Capitalize "gb" to "GB", clarify headers

```suggestion
|          | Cluster Size | GB Over Threshold | Memory Usage per Key    | Primary Node Size | # of Primaries |
| -------- | ------------ | ----------------- | ------------------------------- | ----------------- | -------------- |
| Original | 100 GB       | 95 GB             |  212 B                            | 20 GB             | 5              |
| New      | 58.8 GB      | 0                 |  120 B <br /> (56.6% of original) | 20 GB             | 3              |
```

> So, you can take away two full primaries and whatever associated replicas. As previously mentioned, this is an extreme example, it's unlikely that you have 95% of your data falling outside this threshold. Let's say that you have 60% of your keys that reach above configuration threshold _and_ you can scale back your infrastructure to a smaller-sized machine or instance that has only 16gb of RAM.

Reviewer: ```suggestion
This allows you to remove two primary nodes and their associated replicas. While this is an extreme example, and it is unlikely that 95% of your data falls just outside the threshold, significant savings are still possible in more realistic scenarios. For instance, if 60% of your keys exceed the threshold, you might be able to scale back your infrastructure to smaller machines with 16 GB of RAM.
```

|          | Cluster size | gb above threshold  | Above threshold key size            | Primary size | # of Primaries |
| -------- | ------------ | --------------- | ------------------------------- | ------------ | -------------- |
| Original | 100gb        | 60gb            |  212                            | 20gb         | 5              |

> | New      | 74gb         | 0               |  120 <br /> (56.6% of original) | 14.8gb       | 5              |

Reviewer: ```suggestion
|          | Cluster Size | GB Above Threshold | Memory Usage per Key    | Primary Node Size | # of Primaries |
| -------- | ------------ | ------------------ | ------------------------------- | ----------------- | -------------- |
| Original | 100 GB       | 60 GB              |  212 B                            | 20 GB             | 5              |
| New      | 74 GB        | 0                  |  120 B <br /> (56.6% of original) | 14.8 GB           | 5              |```

> The actual savings here depends on the difference in price for the new vs the old infrastructure, but typically in the cloud or on-prem, there are always break points in price at different specs and this sort of optimization can help you get your cost down when you need just crest over the top of a break point. Of course, in either case, you can always use the new-found space for heretofore uncached data, reduce eviction, or allowing longer TTLs.

Reviewer: ```suggestion
The actual savings depend on the price difference between the new and old infrastructure. In both cloud and on-premises environments, pricing often has distinct breakpoints based on hardware specifications. This optimization can help you drop below a higher pricing tier if you were previously just cresting over it. Alternatively, you can use the reclaimed space to cache previously uncached data, reduce evictions, or allow longer TTLs.
```

> **Wait! How do both a 65 and 64 length value result in a key size of 120?** <br />

> > The eagle-eyed amongst you might have noticed that the original example of `hash1` having a size of 64 bytes and a memory footprint of 120 then in this example we're showing keys that have a value length of 65 that also have 120 bytes memory footprint. 
> > How is that possible?
> > The overhead for hashes with listpack encoding and various sized values isn't entirely linear, it's a stair step. Using the same key pattern and field name, a value length of 49 to 63 was 104 bytes and lengths 64 to 79 was 120 bytes.
> > This explains why, in the original example, a length of 64 increased 15.38% over a length of 63 and yet the encoding was still the same.

Reviewer: ```suggestion
> The eagle-eyed among you might have noticed that while `hash1` (64-byte value) had a memory footprint of 120 bytes, our optimized scenario shows keys with 65-byte values also using 120 bytes. 
> How is that possible?
> Memory usage for hashes with listpack encoding does not increase linearly with value size; it behaves like a stair step. Using the same key pattern and field name, value lengths from 49 to 63 bytes resulted in 104 bytes of memory usage, while lengths from 64 to 79 bytes used 120 bytes.
> This explains why, in the original example, increasing the value length from 63 to 64 bytes resulted in a 15.38% memory increase, even though the encoding remained the same.```

Author: This AI likes use second person plural and anthropomorphize software, which weirdly tracks with how these bots put together language. However, it's criminally confusing in this context.

> All this comes back to understanding that these configurations aren't a silver bullet and throwing wildly higher or lower values at the encoding configuration could result in suboptimal performance or efficiency. YMMV: knowing your data types, sizes, and usage patterns, and (now) how encoding works in Valkey you have the tools to hone your configuration appropriately.

Reviewer: ```suggestion
Ultimately, these configurations are not a silver bullet. Arbitrarily increasing or decreasing these thresholds can result in suboptimal performance or memory efficiency. By understanding your data types, sizes, and usage patterns, along with how Valkey encodings work, you now have the tools to tune your configuration appropriately.
```

## Tell your story

> Did you save a bunch of RAM or opsmaxxing your throughput by tweaking a configuration in Valkey? [Share the knowledge by submitting an issue to write a blog](https://github.com/valkey-io/valkey-io.github.io/issues/new?template=blog_post_template.md) or write a presentation abstract for the upcoming [ValkeyConf](https://events.linuxfoundation.org/valkeyconf/).

Reviewer: ```suggestion
Did you save RAM or maximize your throughput by tweaking a configuration in Valkey? [Share the knowledge by opening an issue to propose a blog post](https://github.com/valkey-io/valkey-io.github.io/issues/new?template=blog_post_template.md) or submitting a presentation abstract for the upcoming [ValkeyConf](https://events.linuxfoundation.org/valkeyconf/).
```

Although -maxxing is linguistically interesting and generally mainstream, probably best to avoid the historical connotation https://en.wikipedia.org/wiki/-maxxing#:~:text=%5B6%5D-,Incel%20adoption,edit,-Main%20article%3A

+++
title = "The secret life of data in Valkey"
date = 2026-06-10
description = "If you’ve used Valkey for very long you’re probably aware of the different primary core data types: string, hashes, lists, sets, sorted sets, and streams. What if that was all a lie."
authors = ["kyledvs"]
[taxonomies]
blog_type = ["Technical Deep Dive"]
[extra]
featured = true
featured_image = "/assets/media/featured/random-05.webp"
+++

If you’ve used Valkey for very long you’re probably aware of the different primary core data types: string, hashes, lists, sets, sorted sets, and streams. Each data type has an accompanying set of commands that manipulate that data type. For example, you use [`HSET`](/commands/hset/) to set a field and value pair to a hash, but you don’t use, for example, [`LPUSH`](/commands/lpush/) to add an element to a list. What if I told you that there is an entire hidden layer that you never see unless you use a specific `OBJECT` command and can’t control unless you start tinkering with the configuration? And that this could have a profound effect on your total usable storage: enough that misconfiguration could mean that you’re wasting much of your infrastructure?

> 

Reviewer 2: A short preview of the blog following the hook would help set the stage for the reader

> If you’ve used Valkey for very long you’re probably aware of the different primary data types in the core: strings, hashes, lists, sets, sorted sets, and streams. Each data type has an accompanying set of commands that manipulate that data type. For example, you use [`HSET`](/commands/hset/) to set a field and value pair to a hash, but you don’t use, for example, HSET to add an element to a list. What if I told you that there is an entire hidden layer that you never see unless you use a specific `OBJECT` command and can’t control unless you start tinkering with the configuration? And that this could have a profound effect on your total usable storage: enough that misconfiguration could mean that you’re wasting much of your infrastructure. Read on to find out how you can understand and tweak the _real_ underlying structures.  

Reviewer 2: still has a negative framing that concerns me. Rather than "misconfiguration could mean waste" why not "careful configuring could mean significant savings"?

```bash
hash-max-listpack-entries 512
hash-max-listpack-value 64
```

This means that any hash with all values less than or equal to 64 in length (`hash-max-listpack-value`) and have less than 512 field-value pairs (`hash-max-listpack-entries`) will be encoded as a 'listpack', everything else as a 'hashtable'. In the original example, key `hash2` had a length of 65, triggering an encoding of 'hashtable', which is much less compact accounting for the 76.67% increase in size. 

There are encoding configurations for:

- Hashes (`hash-max-listpack-entries`, `hash-max-listpack-value`)
- Lists (`list-max-listpack-size`)
- Sets (`set-max-intset-entries`, `set-max-listpack-entries`, `set-max-listpack-value`)
- Sorted Sets (`zset-max-listpack-entries`, `zset-max-listpack-value`)

Streams still use an encoding but there isn’t a threshold because there is only one encoding option. Module types also use encodings but it would vary depending on the logic and/or configuration of the module in question.

### Should you change these settings?

> Maybe. It depends on what you’re optimizing for and what your data looks like. Valkey's default encoding configuration values are fine for many use cases, but if you're strategy is to cost-optimize or maximize your infrastructure, it's worth at least evaluating. 

Reviewer 2: you're strategy -> your strategy

> 

Reviewer 2: clear call to action steps would make this easier for readers. Something like: 

How to evaluate your cluster:

Sample representative keys with OBJECT ENCODING and MEMORY USAGE
Compare value lengths against the default thresholds
If a significant percentage of keys sit just above a threshold, consider raising it
Benchmark throughput before and after — listpack operations are O(n)

Author: This one is tough. I was trying to _not_ give a specific guide to evaluate because one guide isn't likely to be useful. This is an art, not a science. For example, data usage patterns could affect this i.e.: data types not in critical paths could be very optimized while those in critical path should be pushed toward performance, etc.

There are no-brainer situations: let’s say you store data in hashes and they tend to have values in length of 65 bytes, right above the threshold, and you have lower throughput needs. Upping the configuration value for `hash-max-listpack-value` to match a realistic length for your data could deliver huge savings either by lowering the infrastructure requirements or [letting you cache more with the same cluster](https://www.youtube.com/watch?v=cd4-UnU8lWY). 

Contextualizing an extreme example: having a 100gb cluster (5 nodes, 20gb each) packed with 95% of the keys reaching 1 over the `hash-max-listpack-value`. 

> |          | Cluster size | gb wrong-sized  | Wrong-sized key size            | Primary size | # of Primaries |

Reviewer 2: wrong-sized key size confused me. Perhaps "keys using hashtable encoding" or "keys above threshold"

| -------- | ------------ | --------------- | ------------------------------- | ------------ | -------------- |
| Original | 100gb        | 95gb            |  212                            | 20gb        | 5              |
| New      | 58.58gb      | 0               |  120 <br /> (56.6% of original) | 20gb        | 3              |

So, you can take away two full primaries and whatever associated replicas. As previously mentioned, this is an extreme example, it's unlikely that you have 95% of your data falling outside this threshold. Let's say that you have 60% of your keys that reach above the wrong-sized configuration _and_ you can scale back your infrastructure to a smaller-sized machine or instance that has only 16gb of RAM.

|          | Cluster size | gb wrong-sized  | Wrong-sized key size            | Primary size | # of Primaries |
| -------- | ------------ | --------------- | ------------------------------- | ------------ | -------------- |
| Original | 100gb        | 60gb            |  212                            | 20gb         | 5              |
| New      | 74gb         | 0               |  120 <br /> (56.6% of original) | 14.8gb       | 5              |

> The actual savings here depends on the difference in price for the new vs the old infrastructure, but typically in the cloud or on-prem, there are always break points in price at different specs and this sort of optimization can help you get your cost down when you needs just crest over the top of a break point. Of course, in either case, you can always use the new-found space for heretofore uncache data, reduce eviction, or allowing longer TTLs.

Reviewer 2: needs -> need

> The actual savings here depends on the difference in price for the new vs the old infrastructure, but typically in the cloud or on-prem, there are always break points in price at different specs and this sort of optimization can help you get your cost down when you needs just crest over the top of a break point. Of course, in either case, you can always use the new-found space for heretofore uncache data, reduce eviction, or allowing longer TTLs.

Reviewer 2: uncache -> uncached

> **Wait! How do both a 65 and 64 length value result in a key size of 120?** <br />
> The eagle-eyed amongst you might have noticed that the original example of `hash1` having a size of 64 bytes and a memory footprint of 120 then in this example we're showing keys that have a value length of 65 that also have 120 bytes memory footprint. 
> How is that possible?
> The overhead for hashes with listpack encoding and various sized values isn't entirely linear, it's a stair step. Using the same key pattern and field name, a value length of 49 to 63 was 104 bytes and lengths 64 to 79 was 120 bytes.
> This explains why, in the original example, a length of 64 increased 15.38% over a length of 63 and yet the encoding was still the same.

All this comes back to understanding that these configurations aren't a silver bullet and throwing wildly higher or lower values at the encoding configuration could result in suboptimal performance or efficiency. YMMV.

## Tell your story

> Did you save a bunch of RAM or opsmaxxing your throughput by tweaking a configuration in Valkey? [Share the knowledge by submitting an issue to write a blog](https://github.com/valkey-io/valkey-io.github.io/issues/new?template=blog_post_template.md) or write an presentation abstract for the upcoming [ValkeyConf](https://events.linuxfoundation.org/valkeyconf/).

Reviewer 2: an presentation -> a presentation
